components/forwarder/forwarder_test.go

```diff
@@ -0,0 +1,176 @@
+		require.False(t, wasMessageForwarded, "message expected to be not forwarded correctly")
+	})
+
+	t.Run("publish_to_empty_topic", func(t *testing.T) {
+		sentMessage := message.NewMessage(watermill.NewUUID(), message.Payload("message payload"))
+		sentMessage.Metadata = message.Metadata{"key": "value"}
+		err := decoratedPublisherIn.Publish("", sentMessage)
+		require.Error(t, err)
+	})
+}
+
+type PubSubInPublisher struct {
+	message.Publisher
+}
+type PubSubInSubscriber struct {
+	message.Subscriber
+}
+
+type PubSubOutPublisher struct {
+	message.Publisher
+}
+type PubSubOutSubscriber struct {
+	message.Subscriber
+}
+
+func newPubSubIn(logger watermill.LoggerAdapter) (PubSubInPublisher, PubSubInSubscriber) {
+	channelPubSub := gochannel.NewGoChannel(gochannel.Config{}, logger)
+	return PubSubInPublisher{channelPubSub}, PubSubInSubscriber{channelPubSub}
+}
+
+func newPubSubOut(logger watermill.LoggerAdapter) (PubSubOutPublisher, PubSubOutSubscriber) {
+	channelPubSub := gochannel.NewGoChannel(gochannel.Config{}, logger)
+	return PubSubOutPublisher{channelPubSub}, PubSubOutSubscriber{channelPubSub}
+}
+
+func setupForwarder(t *testing.T, ctx context.Context, subscriberIn PubSubInSubscriber, publisherOut PubSubOutPublisher, logger watermill.LoggerAdapter, config forwarder.Config) *forwarder.Forwarder {
+	f, err := forwarder.NewForwarder(subscriberIn, publisherOut, logger, config)
+	require.NoError(t, err)
+
+	go func() {
+		require.NoError(t, f.Run(ctx))
+	}()
+
+	select {
+	case <-f.Running():
+	case <-ctx.Done():
+		t.Error("forwarder not running")
+	}
+
+	return f
+}
+
+func setupMessageForwardedDetectorMiddleware() (message.HandlerMiddleware, <-chan bool) {
+	messageForwardedCh := make(chan bool, 1)
+	messageForwardedDetector := func(handlerFunc message.HandlerFunc) message.HandlerFunc {
+		return func(msg *message.Message) ([]*message.Message, error) {
+			msgs, err := handlerFunc(msg)
+			messageForwardedCh <- err == nil
+
+			// Always return nil as we don't want to nack the message in tests.
+			return msgs, nil
+		}
+	}
+
+	return messageForwardedDetector, messageForwardedCh
+}
+
+func requireFirstForwardingResult(t *testing.T, messageForwardedCh <-chan bool) bool {
+	select {
+	case wasMessageForwarded := <-messageForwardedCh:
+		return wasMessageForwarded
+	case <-time.After(time.Second):
+		t.Fatal("forwarding result not received after 1 sec")
+	}
+
+	return false
+}
+
+func listenOnOutTopic(t *testing.T, ctx context.Context, subscriberOut PubSubOutSubscriber, outTopic string) <-chan *message.Message {
+	messagesCh, err := subscriberOut.Subscribe(ctx, outTopic)
+	require.NoError(t, err)
+
+	return messagesCh
+}
+
+func requireFirstMessage(t *testing.T, expectedMessage *message.Message, ch <-chan *message.Message) {
+	select {
+	case receivedMessage := <-ch:
+		require.NotNil(t, receivedMessage)
+
+		assert.Equal(t, expectedMessage.UUID, receivedMessage.UUID)
```

Comment:
probably it's a better idea to use `message.Equals` here (for the forward compatibility)

Reply:
7154e00